executor/index_merge_reader.go

```diff
@@ -775,18 +789,217 @@ func (w *indexMergeProcessWorker) fetchLoop(ctx context.Context, fetchCh <-chan
+		case workCh <- task:
+			resultCh <- task
+		}
+	}
+}
+
+// For each partition(dynamic mode), a map is used to do intersection. Key of the map is handle, and value is the number of times it occurs.
+// If the value of handle equals the number of partial paths, it should be sent to final_table_scan_worker.
+// To avoid too many goroutines, each intersectionProcessWorker can handle multiple partitions.
+func (w *indexMergeProcessWorker) fetchLoopIntersection(ctx context.Context, fetchCh <-chan *indexMergeTableTask,
+	workCh chan<- *indexMergeTableTask, resultCh chan<- *indexMergeTableTask, finished <-chan struct{}) {
+	defer func() {
+		close(workCh)
+		close(resultCh)
+	}()
+
+	if w.stats != nil {
+		start := time.Now()
+		defer func() {
+			w.stats.IndexMergeProcess += time.Since(start)
+		}()
+	}
+
+	// One goroutine may handle one or multiple partitions.
+	// Max number of partition number is 8192, we use ExecutorConcurrency to avoid too many goroutines.
+	var partCnt int
+	var workerCnt int
+	maxWorkerCnt := w.indexMerge.ctx.GetSessionVars().IndexMergeIntersectionConcurrency()
+	maxChannelSize := atomic.LoadInt32(&LookupTableTaskChannelSize)
+
+	batchSize := w.indexMerge.ctx.GetSessionVars().IndexLookupSize
+	if w.indexMerge.partitionTableMode {
+		partCnt = len(w.indexMerge.prunedPartitions)
+		workerCnt = partCnt
+		if workerCnt > maxWorkerCnt {
+			workerCnt = maxWorkerCnt
+		}
+	} else {
+		partCnt = 1
+		workerCnt = 1
+	}
+	failpoint.Inject("testIndexMergeIntersectionConcurrency", func(val failpoint.Value) {
+		con := val.(int)
+		if con != workerCnt {
+			panic(fmt.Sprintf("unexpected workerCnt, expect %d, got %d", con, workerCnt))
+		}
+	})
+
+	workers := make([]*intersectionProcessWorker, 0, workerCnt)
+	wg := util.WaitGroupWrapper{}
+	errCh := make(chan bool, workerCnt)
+	for i := 0; i < workerCnt; i++ {
+		worker := &intersectionProcessWorker{
+			workerID:            i,
+			handleMapsPerWorker: make(map[int]*kv.MemAwareHandleMap[*int]),
+			workerCh:            make(chan *indexMergeTableTask, maxChannelSize),
+			indexMerge:          w.indexMerge,
+			memTracker:          w.indexMerge.memTracker,
+			batchSize:           batchSize,
+		}
+		wg.RunWithRecover(func() {
+			defer trace.StartRegion(ctx, "IndexMergeIntersectionProcessWorker").End()
+			worker.doIntersectionPerPartition(ctx, workCh, resultCh, finished)
+		}, w.handleLoopFetcherPanic(ctx, resultCh, "IndexMergeIntersectionProcessWorker", errCh))
+		workers = append(workers, worker)
+	}
+	var processWorkerErr bool
+	for task := range fetchCh {
+		select {
+		case workers[task.parTblIdx%workerCnt].workerCh <- task:
+		case processWorkerErr = <-errCh:
+			break
+		}
+	}
+	for _, processWorker := range workers {
+		close(processWorker.workerCh)
+	}
+	wg.Wait()
+	if processWorkerErr {
+		return
+	}
+
+	var totalMemUsage int64
+	for _, processWorker := range workers {
+		totalMemUsage += processWorker.totalMemUsage
+	}
+	w.indexMerge.memTracker.Consume(-(totalMemUsage))
```

Comment:
We can create a tracker for each w.memTrack,  use w.memTracker.Detach() to do this and remove the var totalMemUsage

Reply:
D